README
.SH troff2page

troff2page is a Common Lisp program for making Web
pages from troff documents.  Consult

http://ds26gte.github.io/troff2page/index.html

for its capabilities.

For installation instructions, please see
the file ‘INSTALL’.


history
Last change 2017-08-18

2017 Aug 18

Ported to Lua. Retain 'master' branch for the Lua version, 'lisp'
for the CL.

2016 Feb 26

Added port for Gnu Common Lisp.

2016 Feb 18

html2info functionality incorporated into troff2page.lisp
without need for external shell script.

Corrected os-execute for CLISP and Clozure so they return
proper systat.

2016 Feb 16

html2info -- replace calls to 'nvim -e' with sed, as (1) 'nvim
-e' temporarily takes over screen flashing alarming messages
(even when all commands are marked silent), and (2) Lisp
can't call a shell script containing 'nvim -e' without
hanging.

Macro .troff2info causes troff2page to further convert the HTML doc
to Info doc. This is only done if troff2page can determine
that all aux files have been resolved.

2016 Feb 15

Let *log-stream* be a broadcast stream that combines log
file and stdout. So logging info goes to both console and
file. Have .sy use *log-stream*, so any messages generated
by the system call also get both displayed and saved.

Previously, the user was told to run troff2page again in
order to resolve aux files. Now troff2page will itself take
care to run a second pass. At most two passes are performed
or needed.

CL procedure troff2page:troff2page takes an optional second
argument: If this is non-false, only a single pass is
performed.

pca-t2p-ix.tmac: Avoid requiring a third pass. This uses the
*end-hooks* var in troff2page.lisp.

2016 Feb 14

Script troff2info for converting troff doc to GNU Info doc.

Store log messages (which scroll on stdout) also in
jobname-Z-L.log.

bash/vim script html2info for converting
troff2page-generated HTML doc into a GNU info doc.

2015 Nov 19

.REDIRECT URL

2015 Oct 12

Recognize \p

2015 Sept 28

\*[url ...] and \*[urlh ...] consolidated to just \*[url
...].

\*[url URL]linktext\& becomes <a href="URL">linktext</a>
\*[url]URL\&          becomes <a href="URL">URL</a>

No corresponding definition needed for groff!

Defined string : as abbrev for url.

*

.am doesn't need arg to be already defined

2015 Sept 26

.TH within .TS/.TE

2015 Sept 25

.break .continue

2015 Sept 24

.while

.do

2015 Sept 23

allow parens and spaces within them in arithmetic
expressions. Recognize <? and >?

.nop

-man-style SH and SS defined only for docs known to be -man

2015 Sept 22

\B'...'

Diversion should be retrievable several times.

execute-macro on a diversion should sent it back to input
stream.

Diversion should be callable as string.

2015 Mar 25

Don't do ToC in troff2page.lisp. Move it to a .tmac file.

2015 Mar 19

When first arg of .char is a Unicode escape (\[uXXXX]), make
sure the corresponding Unicode cha gets translated
correctly.

Added string-macro \*[urlh ...].

2014 Aug 22

index.tmac: makeindex has a bug: It won't accept files that
start with a dot! Added workaround.

2013 Sep 2

Allow \h'|...'. Reported by David B. Cortarello.

Support for \f(CB and .fam. By David B. Cortarello.

2011 Nov 25

peek-char in read-possible-troff2page-specific-escape should
not error on eof.  It is possible that a line ends in \
after an .eo.

2011 Nov 23

Don't assume that the .@SH that .SH refers to will be a
(troff2page) request.  It could have been redefined as a
macro.

2011 Oct 21

.LK: produce ToC lines for headers whose levels are <=
\n[GROWPS].

2010 Sep 22

url.tmac: No need for \*[urlh ...].  \*[url ...] can take 1
or 2 args.

.HTML .CDS .CDE; from www.tmac.

2010 Sep 19

SBCL: os-execute should return number indicating exit
status.

do-bye cannot use .blm because input port is already closed.

.nx can take an argument file.

\[uHHHH] is converted to &#xHHHH.  \u[uHHHH] is typically
used by the output of preconv -eutf8.

Mar 13, 2009

.pso should not expand escapes in its arg.  (All escapes, or
only some?)  In any case sed inside a .pso wasn't working
properly because \1 and \) were losing their backslash.

Mar 9, 2009

:external-format arg no longer needed when opening files in
CLISP.

Mar 8, 2009

Make .tm .tmc .sy .pso .writec .write much faster by using
Lisp output primitives instead of going thru the
string-traversing emit procedure.  (These might seem like
unlikely places to optimize, but they are used in the
indexing and eval macros.)

Accept .EQ's optional arguments (indent-style and eqno).

Use *groff-image-options* parameter to influence
magnification and font of .EQ and .PS.

Mar 7, 2009

Allow \\$\\n[numreg] inside macro/string body.

Mar 4, 2009

Call the code file troff2page.lisp, and have troff2page be a
softlink to it.

Mar 2, 2009

Specify charset in HTML file to UTF-8, so UTF-8 chars in
input file are rendered correctly in browsers, instead of
the ugly a-circumflex-euro-trademark trigraphs (?).

Feb 15, 2009

Clean up line processing to accommodate groff's new .lsm macro.

Feb 13, 2009

Line with all spaces (#spaces > 0) should also be treated as blank line
(i.e., .blm will trigger on that line).  Similarly for line
with \" preceded by spaces.  (ToDo: \# .  Is
*keep-newline-p* really needed?)

Feb 11, 2009

.IMG, .PIMG:  alignment argument, which would be the first,
is optional!

Feb 5, 2009

.in \n[.i]

Feb 1, 2009

For man .so, if filename arg in dir above is non-existent,
try in current dir also.  This accommodates .so'ing of aux
files created by troff2page.

Tested on Ubuntu.

Jan 31, 2009

Added support for CMUCL and ECL.

\V

Jan 29, 2009

For man pages, disable toc if only one .SH present.

Arg of .so in man is resolved from one directory above
current file.

Jan 28, 2009

Don't set body max-width in CSS unless user explicitly sets
\n[LL].

.EX/.EE

Move the booklet-style troff2page documentation file to troff2page.ms.
troff2page.1 and eval4troff.tmac.7 are man pages.

Jan 26, 2009

.rn

index.tmac: For nroff, disable index generation and loading
of index (if latter generated anyway via other calls).
Rationale: Any format for cross-refs is clunky; page numbers
may not be useful (esp. if we use a single long page); and

lua/de.lua
-- last modified 2017-08-26

function collect_macro_body(w, ender)
  --print('doing collect_macro_body', w, ender)
  --print('Turn_off_escape_char_p =', Turn_off_escape_char_p)
  local m = {}
  while true do
    if not snoop_char() then
      Exit_status = 'done'
      Macro_spill_over = Control_char .. 'am ' .. w .. ' ' .. ender
      break
    end
    local ln = expand_args(read_troff_line())
    local breakloop
    --print('inside collect_macro_body with', ln)
    flet({
      Current_troff_input = make_bstream {}
    }, function() 
      toss_back_line(ln)
      local c = snoop_char()
      if c == Control_char then
        get_char()
        ignore_spaces()
        local w = read_bare_word()
        --print('checking ->', w, '<-')
        if w == ender then breakloop=true; return end
      end
      table.insert(m, ln)
    end)
    if breakloop then break end
  end
  --print('collect_macro_body returning', table_to_string(m))
  return m
end

function expand_args(s)
  --print('doing expand_args', s)
  if not s then return '' end
  local res = with_output_to_string(function (o)
    flet({
      Current_troff_input = make_bstream{buffer = string_to_table(s)},
      Macro_copy_mode_p = true,
      Outputting_to = 'troff',
      Out = o
    }, function() 
      --print('calling generate_html from expand_args with Out=', Out)
      --print('cti.b =', #(Current_troff_input.buffer))
      generate_html{'break', 'continue', 'ex', 'nx', 'return'}
    end)
  end)
  --print('expand_args retng', res)
  return res
end 

function execute_macro(w)
  --print('doing execute_macro', w)
  local it
  if not w then 
    return
  end
  --
  it = Diversion_table[w]
  if it then
    --print('execing divn')
    read_troff_line()
    local divvalue = retrieve_diversion(it)
    --print('divvalue =', divvalue)
    emit_verbatim(divvalue)
    return
  end
  --
  it = Macro_table[w]
  if it then
    local args = read_args()
    flet({
      Macro_args = args
    }, function()
      table.insert(Macro_args, 1, w)
      execute_macro_body(it)
    end)
    return
  end
  --
  it = String_table[w]
  if it then
    read_troff_line()
    emit_verbatim(it())
    return
  end
  --
  it = Request_table[w] 
  if it then
    --print('calling request', w)
    it()
    return
  end
  --
  read_troff_line()
end

function execute_macro_body(ss)
  --print('doing execute_macro_body', table_to_string(ss))
  flet({
    Macro_spill_over = false
  }, function()
    flet({
      Current_troff_input = make_bstream {}
    }, function()
      local i = #ss
      while true do
        if i<1 then break end
        toss_back_line(ss[i])
        i = i-1
      end
      --print('calling generate_html from execute_macro_body with Out=', Out)
      generate_html {'nx', 'ex'}
    end)
    if Macro_spill_over then
      toss_back_line(Macro_spill_over)
    end
  end)
end

function retrieve_diversion(div)
  local value = div.value
  if not value then 
    value = (div.stream):get_output_stream_string()
    div.value = value
  end
  return value
end
